internal/filesystem/nfs/mountpoint.go

```diff
@@ -0,0 +1,130 @@
+package nfs
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"io/fs"
+	"net"
+	"os"
+
+	"github.com/djdv/go-filesystem-utils/internal/filesystem"
+	"github.com/djdv/go-filesystem-utils/internal/generic"
+	maddrc "github.com/djdv/go-filesystem-utils/internal/multiaddr"
+	"github.com/go-git/go-billy/v5/helper/polyfill"
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/willscott/go-nfs"
+	nfshelper "github.com/willscott/go-nfs/helpers"
+)
+
+type (
+	// Host holds metadata required to host
+	// a file system as an NFS server.
+	Host struct {
+		Maddr multiaddr.Multiaddr `json:"maddr,omitempty"`
+	}
+	// Guest holds metadata required to establish
+	// a client connection to an NFS server.
+	Guest struct {
+		Maddr         multiaddr.Multiaddr `json:"maddr,omitempty"`
+		Hostname      string              `json:"hostname,omitempty"`
+		Dirpath       string              `json:"dirpath,omitempty"`
+		LinkSeparator string              `json:"linkSeparator,omitempty"`
+		LinkLimit     uint                `json:"linkLimit,omitempty"`
+		UID           uint32              `json:"uid,omitempty"`
+		GID           uint32              `json:"gid,omitempty"`
+	}
+)
+
+const (
+	HostID  filesystem.Host = "NFS"
+	GuestID filesystem.ID   = "NFS"
+)
+
+func (*Host) HostID() filesystem.Host { return HostID }
+
+func (nh *Host) UnmarshalJSON(b []byte) error {
+	// multiformats/go-multiaddr issue #100
+	var maddrWorkaround struct {
+		Maddr maddrc.Multiaddr `json:"maddr,omitempty"`
+	}
+	if err := json.Unmarshal(b, &maddrWorkaround); err != nil {
+		return err
+	}
+	nh.Maddr = maddrWorkaround.Maddr.Multiaddr
+	return nil
+}
+
+func (nh *Host) Mount(fsys fs.FS) (io.Closer, error) {
+	listener, err := manet.Listen(nh.Maddr)
+	if err != nil {
+		return nil, err
+	}
+	const cacheLimit = 1024
+	var (
+		netFsys                      = &netFS{fsys: fsys}
+		billyFsys                    = polyfill.New(netFsys)
+		nfsHandler                   = nfshelper.NewNullAuthHandler(billyFsys)
+		cachedHandler                = nfshelper.NewCachingHandler(nfsHandler, cacheLimit)
+		goListener                   = manet.NetListener(listener)
+		errsCh                       = make(chan error, 1)
+		closerFn      generic.Closer = func() error {
+			if err := listener.Close(); err != nil {
+				return err
+			}
+			if err := <-errsCh; !errors.Is(err, net.ErrClosed) {
+				return err
+			}
+			return nil
+		}
+	)
+	// The NFS library has verbose logging by default.
+	// If the operator has not specified a log level,
+	// override the library's default level.
+	// (Primarily to suppress `ENOENT` errors in the console.)
+	const nfslibEnvKey = "LOG_LEVEL"
+	if _, set := os.LookupEnv(nfslibEnvKey); !set {
+		nfs.Log.SetLevel(nfs.PanicLevel)
+	}
+	go func() { errsCh <- nfs.Serve(goListener, cachedHandler) }()
+	return closerFn, nil
+}
+
+func (*Guest) GuestID() filesystem.ID { return GuestID }
```

Comment:
this means you're only able to have one mount of this at a time because it'll always have the same global filesystem ID?

Reply:
Nah. It's just the name of the file system.
That interface method is probably poorly name. It used to be `FSID` iirc. Maybe it should be `.Type` or something else?

I forget all the places it's used, but I think the mount table uses it. The mount commands might use it, but I think not for "reasons".

Check this out, if we start the daemon with an exposed socket, it acts as a 9P server.
We can then instantiate a bunch of mount points, then as a 9P client, see what the actual mount table looks like.
Which shows those names being used in both the paths and mountpoint tags.
```
Windows:
fs daemon -api-server /ip4/192.168.1.40/tcp/564

fs mount nfs keyfs -api-server /ip4/192.168.1.40/tcp/564 /ip4/192.168.1.40/tcp/2050
fs mount nfs pinfs -api-server /ip4/192.168.1.40/tcp/564 /ip4/192.168.1.40/tcp/2051

fs mount nfs pinfs -api-server /ip4/192.168.1.40/tcp/564 /ip4/192.168.1.40/tcp/2049
fs.exe mount fuse nfs -api-server /ip4/192.168.1.40/tcp/564 -nfs-server /ip4/192.168.1.40/tcp/2049 I:

Linux:
mount -t 9p 192.168.1.40 /mnt/9 -o trans=tcp,port=564
tree /mnt/9
```
![Screenshot 2024-02-06 105930](https://github.com/djdv/go-filesystem-utils/assets/13862850/0d511200-60e8-44af-8d42-716a9826197d)